codex/blockexchange/engine/engine.nim

```diff
@@ -279,50 +282,107 @@ proc scheduleTasks(b: BlockExcEngine, blocks: seq[bt.Block]) {.async.} =
 
           break # do next peer
 
-proc resolveBlocks*(b: BlockExcEngine, blocks: seq[bt.Block]) {.async.} =
-  trace "Resolving blocks", blocks = blocks.len
+proc resolveBlocks*(b: BlockExcEngine, blocksDelivery: seq[BlockDelivery]) {.async.} =
+  trace "Resolving blocks", blocks = blocksDelivery.len
 
-  b.pendingBlocks.resolve(blocks)
-  await b.scheduleTasks(blocks)
-  b.discovery.queueProvideBlocksReq(blocks.mapIt( it.cid ))
+  b.pendingBlocks.resolve(blocksDelivery)
+  await b.scheduleTasks(blocksDelivery)
+  var cids = initHashSet[Cid]()
+  for bd in blocksDelivery:
+    cids.incl(bd.blk.cid)
+    if bd.address.leaf:
+      cids.incl(bd.address.treeCid)
+  b.discovery.queueProvideBlocksReq(cids.toSeq)
+
+proc resolveBlocks*(b: BlockExcEngine, blocks: seq[Block]) {.async.} =
+  await b.resolveBlocks(blocks.mapIt(BlockDelivery(blk: it, address: BlockAddress(leaf: false, cid: it.cid))))
 
 proc payForBlocks(engine: BlockExcEngine,
                   peer: BlockExcPeerCtx,
-                  blocks: seq[bt.Block]) {.async.} =
-  trace "Paying for blocks", blocks = blocks.len
+                  blocksDelivery: seq[BlockDelivery]) {.async.} =
+  trace "Paying for blocks", len = blocksDelivery.len
 
   let
     sendPayment = engine.network.request.sendPayment
-    price = peer.price(blocks.mapIt(it.cid))
+    price = peer.price(blocksDelivery.mapIt(it.address))
 
   if payment =? engine.wallet.pay(peer, price):
     trace "Sending payment for blocks", price
     await sendPayment(peer.id, payment)
 
-proc blocksHandler*(
+proc validateBlockDelivery(
+  b: BlockExcEngine,
+  bd: BlockDelivery
+): ?!void =
+  if bd.address notin b.pendingBlocks:
+    return failure("Received block is not currently a pending block")
```

Comment:
@dryajov this is a semantic change. Previously we stored any blocks delivered to us. Now I added this validation step to check if we were awaiting for a delivered block in the first place. If not we're discarding it. 

I added it to prevent someone filling up our disk with some random blocks, that we didn't ask for.

Reply:
Yeah, good check